File: core/GreatestCommonDivisorOfStrings/GreatestCommonDivisorOfStrings.py
def gcdOfStrings(str1: str, str2: str) -> str:
    length1, length2 = len(str1), len(str2)

    def is_div_str(cur_str, length):
        temp_str = cur_str[:length]
        temp_length = length
        while temp_length <= len(cur_str):
            if temp_str != cur_str[temp_length - length:temp_length]:
                return False
            temp_length += length

        return True

    point = min(length1, length2)
    while point > 0:
        if length1 % point == 0 and length2 % point == 0 and\
                str1[:point] == str2[:point] and is_div_str(str1, point) and is_div_str(str2, point):
            return str1[:point]
        point -= 1

    return ""


if __name__ == '__main__':
    print(gcdOfStrings(str1="ABCABC", str2="ABC"))

File: core/GreatestCommonDivisorOfStrings/GreatestCommonDivisorOfStrings.java
package com.dbc;

public class GreatestCommonDivisorOfStrings {
    public String gcdOfStrings(String str1, String str2) {
        int candi = getGcd(str1.length(), str2.length());
        if ((str1 + str2).equals(str2 + str1)) {
            return str1.substring(0, candi);
        }
        return "";
    }

    private int getGcd(int a, int b) {
        int remain = a % b;
        while (remain != 0) {
            a = b;
            b = remain;
            remain = a % b;
        }
        return b;
    }
}
